src/libraries/Microsoft.Extensions.Logging.Abstractions/src/IBufferedLogRecordMetadata.cs

```diff
@@ -0,0 +1,39 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Diagnostics;
+
+namespace Microsoft.Extensions.Logging
+{
+    /// <summary>
+    /// Holds metadata for a buffered log record.
+    /// </summary>
+    /// <remarks>
+    /// If the logging infrastructure decides to buffer log records in memory, it will ensure that the
+    /// <c>TState</c> value delivered to logging providers implements this interface. The logging providers
+    /// can then use the metadata to augment the log records they emit.
+    /// </remarks>    
+    public interface IBufferedLogRecordMetadata
```

Comment:
>IBufferedLogRecordMetadata [](http://example.com/codeflow?start=21&length=26)

do we think this interface will need to be expanded in the future? I am asking because we still need to support .NET Framework which doesn't support [default interface methods](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/default-interface-methods). 

If so, we may have the interface like 

```C#
    public interface IBufferedLogRecordMetadata
    {
        public BufferedLogRecordMetadata { get; }
    }
```

and 

```C#
public sealed class BufferedLogRecordMetadata
{
        /// <summary>
        /// Gets the time when the log record was recorded.
        /// </summary>
        public DateTimeOffset CreationTime { get; init; }

        /// <summary>
        /// Gets the ID of the thread that created the log record.
        /// </summary>
        public int? ManagedThreadId { get; init; }

        /// <summary>
        /// Gets the ID of the span in effect when the log record was created.
        /// </summary>
        public ActivitySpanId? SpanId { get; init; }

        /// <summary>
        /// Gets the ID of the trace in effect when the log record was created.
        /// </summary>
        public ActivityTraceId? TraceId { get; init; }

}
```

Reply:
> do we think this interface will need to be expanded in the future?

I think we should certainly leave that possibility open as long as it doesn't impose a high cost (I don't believe it would)

Reply:
> public sealed class BufferedLogRecordMetadata

Perhaps `public sealed struct BufferedLogRecordMetadata` so we aren't allocating a new object to transfer 4 fields?
Alternatively if we have a new field we want to share in the future we could define a new interface IBufferLogRecordXYZMetadata and the TState implements both of them. Given that there will probably be < 10 pieces of code in the world that interact with this type it wouldn't bother me at all to use a quick and dirty approach.

Reply:
I prefer the `struct` idea than introducing more interfaces in the future. Additions to the `struct` will be much simpler.

Reply:
@noahfalk did you intent to use `readonly` modifier? Because we can't use `sealed` for structs

Reply:
Make it into a struct, returned as an out parameter.

Reply:
> @noahfalk did you intend to use readonly modifier? Because we can't use sealed for structs

Ignore the sealed, I just copy/pasted and changed 'class' to 'struct' without paying attention to the other modifiers : )